### PR TITLE
Adding a CNN encoder

### PR DESCRIPTION
**Pull request by matt-gardner**

### CODE REVIEW

File: allennlp/modules/seq2vec_encoders/cnn_encoder.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    def forward(self, tokens: torch.Tensor):  # pylint: disable=arguments-differ
+        # Our input is expected to have shape `(batch_size, num_tokens, embedding_dim)`.  The
+        # convolution layers expect input of shape `(batch_size, in_channels, sequence_length)`,
+        # where the conv layer `in_channels` is our `embedding_dim`.  We thus need to transpose the
+        # tensor first.
+        tokens = torch.transpose(tokens, 1, 2)
+
+        # Each convolution layer returns output of size `(batch_size, num_filters, pool_length)`,
+        # where `pool_length = num_tokens - ngram_size + 1`.  We then do an activation function
+        # (currently hard-coded to RELU), then do max pooling over each
```
Comment (DeNeutoy) [inline allennlp/modules/seq2vec_encoders/cnn_encoder.py]: Line wrapping is weird here.

File: allennlp/modules/seq2vec_encoders/cnn_encoder.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        # filter for the whole input sequence.  Because our max pooling is simple, we just use
+        # `torch.max`.  The resultant tensor of has shape `(batch_size, num_conv_layers *
+        # num_filters)`, which then gets projected using the projection layer, if requested.
+        # TODO(mattg): allow specifying the activation function to use here.
```
Comment (DeNeutoy) [inline allennlp/modules/seq2vec_encoders/cnn_encoder.py]: Why is this a TODO? Can't you just add it to `__init__`?

File: allennlp/modules/seq2vec_encoders/cnn_encoder.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        activation = torch.nn.functional.relu
+
+        filter_outputs = [torch.max(activation(convolution_layer(tokens), inplace=True),
+                                    dim=2)[0].squeeze(dim=2)  # not really sure why max isn't squeezing...
```
Comment (DeNeutoy) [inline allennlp/modules/seq2vec_encoders/cnn_encoder.py]: Looks like we can get this if we upgrade versions? http://pytorch.org/docs/master/torch.html#torch.max
I doubt you'd be able to do the max inplace and squeeze it simultaneously though.

File: allennlp/modules/seq2vec_encoders/cnn_encoder.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                          for convolution_layer in self._convolution_layers]
+
+        # Now we have a list of `num_conv_layers` tensors of shape `(batch_size, num_filters)`.
+        # Concatenating them gives us a tensor of shape
```
Comment (DeNeutoy) [inline allennlp/modules/seq2vec_encoders/cnn_encoder.py]: Inconsistent line wrapping here.

File: allennlp/modules/seq2vec_encoders/cnn_encoder.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    the output after maxpooling is ``len(ngram_filter_sizes) * num_filters``.  This then gets
+    (optionally) projected down to a lower dimensional output, specified by ``output_dim``.
+
+    We then use a fully connected layer to project in back to the desired output_dim.  For more
+    details, refer to "A Sensitivity Analysis of (and Practitioners’ Guide to) Convolutional Neural
+    Networks for Sentence Classification", Zhang and Wallace 2016, particularly Figure 1.
+
+    Parameters
+    ----------
+    embedding_dim : ``int``
```
Comment (DeNeutoy) [inline allennlp/modules/seq2vec_encoders/cnn_encoder.py]: Are we doing ``int`` or just "int"? I've just been using "int", but you have both here.

File: allennlp/modules/seq2vec_encoders/cnn_encoder.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+from torch.nn import Conv1d, Linear
+
+from allennlp.common import Params
+from allennlp.experiments import Registry
+from allennlp.modules import Seq2VecEncoder
+
+@Registry.register_seq2vec_encoder("cnn")
+class CnnEncoder(Seq2VecEncoder):
+    """
+    CnnEncoder is a combination of multiple convolution layers and max pooling layers.  As a
```
Comment (DeNeutoy) [inline allennlp/modules/seq2vec_encoders/cnn_encoder.py]: "A CnnEncoder"

File: allennlp/modules/seq2vec_encoders/cnn_encoder.py (hunk 3 above)
Comment (DeNeutoy) [inline allennlp/modules/seq2vec_encoders/cnn_encoder.py]: In the meantime this might be more readable as `activation(convolution_layer(tokens.max(dim=2)[0].squeeze(dim=2)`.

File: allennlp/modules/seq2vec_encoders/cnn_encoder.py (hunk 5 above)
Comment (matt-gardner) [inline allennlp/modules/seq2vec_encoders/cnn_encoder.py]: This really depends on how we're building our docs.  If we're doing the same as what we did with DeepQA, using double back ticks here is preferred.

File: allennlp/modules/seq2vec_encoders/cnn_encoder.py (hunk 2 above)
Comment (matt-gardner) [inline allennlp/modules/seq2vec_encoders/cnn_encoder.py]: Because we don't have a list of activation functions, or an obvious type to take, and I don't know enough yet about how pytorch does activations...

File: allennlp/modules/seq2vec_encoders/cnn_encoder.py (hunk 3 above)
Comment (matt-gardner) [inline allennlp/modules/seq2vec_encoders/cnn_encoder.py]: The inplace argument is for the activation, not the max.  I'm not sure what you're suggesting would be more readable, as there are unmatched parentheses, and you're calling max on tokens, not the result of the activation function.

File: allennlp/modules/seq2vec_encoders/cnn_encoder.py (hunk 4 above)
Comment (matt-gardner) [inline allennlp/modules/seq2vec_encoders/cnn_encoder.py]: This one was intentional (the other one wasn't) - I thought it'd be easier to read if the shape was on one line.  If you don't think so, I can fix it.

File: allennlp/modules/seq2vec_encoders/cnn_encoder.py (hunk 6 above)
Comment (matt-gardner) [inline allennlp/modules/seq2vec_encoders/cnn_encoder.py]: Done.

File: allennlp/modules/seq2vec_encoders/cnn_encoder.py (hunk 3 above)
Comment (DeNeutoy) [inline allennlp/modules/seq2vec_encoders/cnn_encoder.py]: Sorry, I meant: `activation(convolution_layer(tokens)).max(dim=2)[0].squeeze(dim=2)`.
